spec/integration/application/agent_spec.rb

```diff
@@ -348,4 +348,56 @@
       end
     end
   end
+
+  context 'multiple agents running' do
+    before(:each) do
+      File.delete(Puppet[:agent_catalog_run_lockfile]) if File.file?(Puppet[:agent_catalog_run_lockfile])
+    end
+
+    it "exits if an agent is already running" do
+      server.start_server do |port|
+        path = Puppet[:agent_catalog_run_lockfile]
+        Puppet[:masterport] = port
+
+        t1 = Thread.new {
+          %x{ruby -e "File.write('#{path}', Process.pid); sleep(3); 'puppet keyword needed here'"}
+        }
+
+       sleep 0.5 # so that the file gets created
+
+        expect {
+          expect {
+            agent.command_line.args << '--test'
+            agent.run
+          }.to exit_with(1)
+        }.to output(/Run of Puppet configuration client already in progress; skipping/).to_stdout
+        
+        t1.kill # kill thread so we don't wait too much
+      end
+    end
+
+    it "waits for other agent run to finish before starting" do
+      server.start_server do |port|
+        path = Puppet[:agent_catalog_run_lockfile]
+        Puppet[:masterport] = port
+        Puppet[:waitforlock] = 1
+
+       t = Thread.new {
+          %x{ruby -e "File.write('#{path}', Process.pid); sleep(4); 'puppet keyword needed here'"}
+       }
+
+       sleep 0.5 # so that the file gets created
```

Comment:
Looks like appveyor failed on some nodes due to the above race conditions.

```
  2) puppet agent multiple agents running waits for other agent run to finish before starting
     Failure/Error:
         expect {
         expect {
           agent.command_line.args << '--test'
           agent.run
         }.to exit_with(0)
       }.to output(/Info: Will try again in #{Puppet[:waitforlock]} seconds./).to_stdout
       expected block to output /Info: Will try again in 1 seconds./ to stdout, but output "\e[0;32mInfo: Using configured environment 'production'\e[0m\n\e[0;32mInfo: Retrieving pluginfacts\e...s20200416-2984-4a1yq8/var/state/state.yaml\e[0m\n\e[mNotice: Applied catalog in 0.03 seconds\e[0m\n"
       Diff:
       @@ -1,2 +1,7 @@
       -/Info: Will try again in 1 seconds./
       +Info: Using configured environment 'production'
       +Info: Retrieving pluginfacts
       +Info: Retrieving plugin
       +Info: Caching catalog for 127.0.0.1
       +Info: Creating state file C:/Users/appveyor/AppData/Local/Temp/rspecrun20200416-2984-v8k5ur/tmp_settings20200416-2984-4a1yq8/var/state/state.yaml
       +Notice: Applied catalog in 0.03 seconds
     # ./spec/integration/application/agent_spec.rb:369:in `block (4 levels) in <top (required)>'
     # ./spec/lib/puppet_spec/puppetserver.rb:95:in `start_server'
```

You could have the main thread poll for the file to be created, something like:

```ruby
until File.exist?(path) && File.size(path) > 0 do
  sleep 0.1
end
```